crates/subspace-solving/src/lib.rs

```diff
@@ -24,11 +24,30 @@ mod codec;
 pub use codec::SubspaceCodec;
 use schnorrkel::SignatureResult;
 use sha2::{Digest, Sha256};
-use subspace_core_primitives::{crypto, LocalChallenge, Piece, Randomness, Salt, Tag, TAG_SIZE};
+use subspace_core_primitives::{
+    crypto, LocalChallenge, Piece, PieceIndexHash, PublicKey, Randomness, Salt, Tag, TAG_SIZE,
+};
 
 /// Signing context used for creating solution signatures by farmer
 pub const SOLUTION_SIGNING_CONTEXT: &[u8] = b"farmer_solution";
 
+uint::construct_uint! {
+    /// Distance to piece index hash from farmer identity
+    pub struct PieceDistance(4);
+}
+
+impl PieceDistance {
+    /// Calculates the xor distance metric between piece index hash and farmer address.
+    pub fn xor_distance(PieceIndexHash(piece): &PieceIndexHash, address: &PublicKey) -> Self {
+        Self::from_big_endian(piece) ^ Self::from_big_endian(address.as_ref())
+    }
+
+    /// Convert piece distance to big endian bytes
+    pub fn as_bytes(self) -> [u8; 32] {
+        self.into()
+    }
+}
```

Comment:
True, but wouldn't it be simpler to think about distance as just u256 number? For tags we can just use regular xor.

Reply:
It is, but my point was that this function is useful in several instances, so I'd rather generalize it and then probably move to `subspace-core-primitives`. Since the size is just U256 we can just read 2 u128 numbers from it and will not need any third-party libraries for that. Just feels like overengineering the way it is right now.

Reply:
Yeah, but you can make a `xor_distance` using standart trait `Xor` from `std::ops` if we want to, and it can be reused.

Reply:
That is what above code uses, it should even compile to something efficient with SIMD instructions.

Reply:
True, but it would make it harder to reason some things like comparison of distance or other things.
Xor distance itself is just xor on  data with same type. Maybe we should just remove this function then? Or just force types for constructing distance

Reply:
I was thinking about a set of functions for these kinds of operations. Or maybe just pairs for 64-bit and another for 256-bit (instead of arbitrary size), but for general purpose arrays, you can turn any of those data structures into it.

Reply:
Yeah, right, but this is not just an operation on some random data. It is always either xor distance from local challenge target to tag, or from identity to piece index hash. It just doesn't seem right to make those using generics, as it won't be reused much and it is not that many code anyways.

Speaking of operations, with `uint` crate they get generated automatically and all the code becomes super readable, like here:
https://github.com/subspace/subspace/pull/304/files#diff-af6e7430bf613b8abd27401b30a3bc6f88d512220979d7cd8e5e625957ccb7b0R187

Reply:
I think we can merge this as is and figure out what to do with it, if anything, later. This should work fine for now.